llvm/lib/Transforms/Scalar/Reassociate.cpp

```diff
@@ -648,6 +651,9 @@ static bool LinearizeExprTree(Instruction *I,
     // Ensure the leaf is only output once.
     It->second = 0;
     Ops.push_back(std::make_pair(V, Weight));
+    if (Opcode == Instruction::Add && Flags.AllKnownNonNegative) {
```

Comment:
```suggestion
    if (Opcode == Instruction::Add && Flags.AllKnownNonNegative && Flags.HasNSW) {
```
Can we avoid computing this when `HasNSW` is false?

Reply:
That makes sense, just that there could be a case where `Flags.AllKnownNonNegative` would be true even if one of the operands is actually negative. (Just because HasNSW is false.

Perhaps a comment explaining that `Flags.AllKnownNonNegative` cannot be used independent of `Flags.HasNSW`